test/weiner_convolve_test.cc

```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright(c) 2019 Netflix, Inc.
+ * SPDX - License - Identifier: BSD - 2 - Clause - Patent
+ */
+
+/*
+ * Copyright (c) 2017, Alliance for Open Media. All rights reserved
+ *
+ * This source code is subject to the terms of the BSD 2 Clause License and
+ * the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+ * was not distributed with this source code in the LICENSE file, you can
+ * obtain it at www.aomedia.org/license/software. If the Alliance for Open
+ * Media Patent License 1.0 was not distributed with this source code in the
+ * PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+ */
+
+/******************************************************************************
+ * @file weiner_convolve_test.cc
+ *
+ * @brief Unit test of weiner convolbe add source:
+ * - av1_wiener_convolve_add_src_avx2
+ * - av1_highbd_wiener_convolve_add_src_avx2
+ *
+ * @author Cidana-Edmond
+ *
+ ******************************************************************************/
+
+#include "gtest/gtest.h"
+// workaround to eliminate the compiling warning on linux
+// The macro will conflict with definition in gtest.h
+#ifdef __USE_GNU
+#undef __USE_GNU  // defined in EbThreads.h
+#endif
+#ifdef _GNU_SOURCE
+#undef _GNU_SOURCE  // defined in EbThreads.h
+#endif
+#include "aom_dsp_rtcd.h"
+#include "EbDefinitions.h"
+#include "EbRestoration.h"
+#include "convolve.h"
+#include "random.h"
+#include "util.h"
+
+/**
+ * @brief Unit test of weiner convolbe add source:
+ * - av1_wiener_convolve_add_src_avx2
+ * - av1_highbd_wiener_convolve_add_src_avx2
+ *
+ * Test strategy:
+ * Verify this assembly code by comparing with reference c implementation.
+ * Feed the same data and check test output and reference output.
+ * Define a templete class to handle the common process, and
+ * declare sub class to handle different bitdepth and function types.
+ *
+ * Expected result:
+ * Output from assemble functions should be the same with output from c.
+ *
+ * Test coverage:
+ * Test cases:
+ * BitDepth: 8bit, 10bit and 12bit
+ */
+
+namespace {
+using std::make_tuple;
+using svt_av1_test_tool::SVTRandom;
+
+class SVTRandomHelper {
```

Comment:
Seems like the main body of this class is to generate kernels with different type, should we put this into base class instead of multiple class inherit?

Reply:
merged